Day 6. Homework

eclimov opened the pull request:
The following points have been implemented (+ some improvements):
➢ On list jobs page show only active jobs grouped by category
➢ Show only 10 jobs per category on this page
➢ Expired jobs should not be accessible by link
➢ Install StofDoctrineExtensionsBundle and setup slug for category
(don’t forget about migration)
➢ Create action with route /category/{slug} where will be displayed
all jobs by specific category
➢ Link this page with the page with all categories

eclimov committed:
- Installed "StofDoctrineExtensionsBundle"
- Added global twig variable "jobs_show_limit_main_page"
- Added "CategoryController"
- Updated "JobController":
now getting a list of jobs for main page through Category entity;
some "best practices" changes applied
- Not showing expired or not activated jobs on main page anymore
- Only active jobs are accessible through direct url now
- Updated "JobFixtures": more random jobs are being created for better debug experience (number of categories remains the same)
- Updated "Category" entity:
added "slug" feature;
added "getActiveJobs" method
- Updated "Job" entity:
added "isActive()" method (note: a job is considered active, when it is activated and not expired);
updated "prePersist()" method, - "expiresAt" field update doesn't overwrite fixtures anymore
- Added and executed migrations
- Added "CategoryRepository" with "findCategoriesWithActiveJobs()" method
- Added "findActiveJob()" method to "JobRepository"
- Added "category_job_list" template: it outputs all jobs of a certain category, without conditions
- Updated "list" template:
jobs are now grouped by category on main page;
improved layout

gregurco commented on config/packages/twig.yaml:
```diff
@@ -2,3 +2,5 @@ twig:
     paths: ['%kernel.project_dir%/templates']
     debug: '%kernel.debug%'
     strict_variables: '%kernel.debug%'
+    globals:
```
notice that in all yaml files you have 4 spaces on one level, so finally you should have:
```
twig:
    paths: ['%kernel.project_dir%/templates']
    debug: '%kernel.debug%'
    strict_variables: '%kernel.debug%'
    globals:
        jobs_show_limit_main_page: 10
```
add 2 more spaces before  `jobs_show_limit_main_page`

gregurco replied:
also fix spaces in `config/packages/stof_doctrine_extensions.yaml`

gregurco commented on src/Controller/CategoryController.php:
```diff
@@ -0,0 +1,29 @@
+class CategoryController extends AbstractController
+{
+    /**
+     * @Route("/category/{slug}", name="category_job_list")
```
don't repeat `category` word in route name here. Finally you will have `category.category_job_list`, better to be `category.job_list`

gregurco commented on src/Controller/CategoryController.php:
```diff
@@ -0,0 +1,29 @@
+    {
+        return $this->render('category/category_job_list.html.twig', [
+            'category' => $category,
+            'jobs' => $category->getJobs(),
```
in twig you will have category object with jobs two, why do you need to separate them in two variables and to send in twig job with categories and categories?

eclimov replied:
It's for page name in "title' block. I will update this, so only category name will be passed to the template, instead of whole object.

gregurco commented on src/Controller/JobController.php:
```diff
@@ -2,36 +2,44 @@
 
         return $this->render('job/list.html.twig', [
-           'jobs' => $jobs
+           'categories' => $categories
```
add `,` at last element in array

gregurco commented on src/Entity/Job.php:
```diff
@@ -430,7 +438,10 @@ public function setCategory(?Category $category): Job
-        $this->expiresAt = new \DateTime('+30 days');
         $this->createdAt = new \DateTime();
+
+        if(!$this->expiresAt){
```
code style: add space before if: 
```php
if (condition) {
    // ...
}
```

gregurco commented on templates/job/list.html.twig:
```diff
@@ -3,22 +3,31 @@
+        {% for category in categories %}
+            <div id="category_section" class="mt-3 mb-4">
+                <h4>
+                    <a href="{{ url('category.category_job_list', {slug: category.getSlug()}) }}">{{ category.getName() }}</a>
```
better `category.slug`

eclimov committed:
- Fixed indentations in "yaml" files, according to coding style
- Updated template in "category" directory: it doesn't contain word "category" anymore
- Not repeating entity's name in route's name anymore
- Not passing whole "Category" object to the template anymore, - passing categoryName, instead
- Added missing trailing commas to arrays
- Added spaces to "if" instruction, according to coding style
- Slightly updated "job.list" template, according to best practices

eclimov commented:
Fixed
